controller/src/main/java/io/pravega/controller/task/Stream/StreamTransactionMetadataTasks.java

```diff
@@ -198,80 +254,330 @@ public Void initializeStreamWriters(final String commitStreamName, final EventSt
-                                                                                           final long scaleGracePeriod,
-                                                                                           final OperationContext ctx) {
-        UUID txnId = UUID.randomUUID();
-        return streamMetadataStore.createTransaction(scope, stream, txnId, lease, maxExecutionPeriod,
-                scaleGracePeriod, ctx, executor)
-                .thenComposeAsync(txData -> streamMetadataStore.getActiveSegments(scope, stream, txData.getEpoch(), ctx, executor)
-                        .thenComposeAsync(segmentNumbers -> notifyTxnCreation(scope, stream, segmentNumbers, txnId)
-                                .thenComposeAsync(v -> FutureHelpers.allOfWithResults(segmentNumbers.stream()
-                                        .map(s -> streamMetadataStore.getSegment(scope, stream, s, ctx, executor))
-                                        .collect(Collectors.toList()))
-                                        .thenApply(segments -> new ImmutablePair<>(txData, segments)), executor), executor), executor);
+    /**
+     * Ping a txn thereby updating its timeout to current time + lease.
+     *
+     * Post-condition:
+     * 1. If ping request completes successfully, then
+     *     (a) txn timeout is set to lease + current time in timeout service,
+     *     (b) txn version in timeout service equals version of txn node in store,
+     *     (c) if txn's timeout was not previously tracked in timeout service of current process,
+     *     then version of txn node in store is updated, thus fencing out other processes tracking timeout for this txn,
+     *     (d) txn is present in the host-txn index of current host,
+     *
+     * 2. If process fails before responding to the client, then since txn is present in the host-txn index,
+     * some other controller process shall abort the txn after maxLeaseValue
+     *
+     * Store read/update operation is not invoked on receiving ping request for a txn that is being tracked in the
+     * timeout service. Otherwise, if the txn is not being tracked in the timeout service, txn node is read from
+     * the store and updated.
+     *
+     * @param scope      scope name.
+     * @param stream     stream name.
+     * @param txnId      txn id.
+     * @param lease      txn lease.
+     * @param ctx        context.
+     * @return           ping status.
+     */
+    CompletableFuture<PingTxnStatus> pingTxnBody(final String scope,
+                                                 final String stream,
+                                                 final UUID txnId,
+                                                 final long lease,
+                                                 final OperationContext ctx) {
+        if (!timeoutService.isRunning()) {
+            return CompletableFuture.completedFuture(createStatus(Status.DISCONNECTED));
```

Comment:
If timeoutService is disconnected, client gets an error message to that effect. On receiving It needs to send ping to some other controller instance to keep the transaction alive.
Currently, client does not send ping messages on behalf of the transactions it creates. There is an issue to resolve that problem.